### PR TITLE
Installation issues

### PR DESCRIPTION
**Pull request by chrisallenlane**
Resolves the following:

- #351 (use of `sudo` when installing)
- #420 (failure to install on Windows)
- #431 (failure to install on MacOS)

Application now relies on `appdirs` module to identify the appropriate
locations for storing configuration and data, both during installation
and runtime.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -1,10 +1,23 @@
...
-# determine the directory in which to install system-wide cheatsheets
-# KLUDGE: It would be better to read `/usr/share/cheat` from `config/cheat`
-# rather than hard-coding it here
-cheat_path = os.environ.get('CHEAT_PATH') or '/usr/share/cheat'
+# install appdirs if it cannot be imported
+try:
+    import appdirs
+except ImportError:
+    import pip
+    pip.main(['install', 'appdirs'])
```
Comment (m3nu) [inline setup.py]: This failed with `AttributeError: module 'pip' has no attribute 'main'`

File: setup.py [hunk 2]
```diff
@@ -1,10 +1,23 @@
...
+    import appdirs
+
+# determine the path in which to install the cheatsheets included with the
+# package
+cheat_path = os.environ.get('CHEAT_PATH') or \
+                appdirs.user_data_dir('cheat', 'cheat')
+
```
Comment (m3nu) [inline setup.py]: This is great.

File: README.md [hunk 3]
```diff
@@ -50,9 +50,12 @@ Installing
 It is recommended to install `cheat` with `pip`:
 
 ```sh
-[sudo] pip install cheat
+pip install cheat --user
 ```
 
+(You must ensure that the `Location` identified by `pip show cheat` exists on
+your `PATH`.)
+
```
Comment (m3nu) [inline README.md]: Line 68 will need updating as well. It references `~/.cheat`

**Comment by equal-l2** [comment]
Hello. Is this still WIP? 
#431 blocks an update on Homebrew : https://github.com/Homebrew/homebrew-core/pull/37081

**Comment by chenrui333** [comment]
Any update on this issue? Thanks!

**Comment by chrisallenlane** [comment]
Hi, all

Sorry - I had another project completely take over my life for the past few months. I should be free to return my attention to `cheat` by next week (June). Sorry about the delays!

**Comment by ryanjbonnell** [comment]
I'm a proud supporter of Cheat and a long-time Mac user myself, so it would be nice to resolve these installation issues before the next major version of macOS comes out this Fall.

With this in mind, I tested this branch on macOS 10.14 "Mojave" and was able to install Cheat 2.5.1 successfully, albeit with the following caveats:

1. The current version of macOS ships with Python 2.7 installed — but without `pip` — so I had to first [install pip on macOS](https://stackoverflow.com/a/38043109/245264):

`$ python --version && sudo easy_install pip`

Once I had pip, I could attempt installing Cheat from this `installation-issues` branch.

2. Attempting to install Cheat without sudo fails:

`$ pip install https://github.com/cheat/cheat/archive/installation-issues.zip`

Re-running the same command as above with sudo allows the installation to succeed, as it allows `setup.py` to run without the permission denied errors in `/Library/Python/2.7/site-packages/cheat/configuration.py`.

Despite all of this, installing via Homebrew is likely far more preferable for Mac users of all types — especially given the need for pip's sudo access.

> _Ironically, Homebrew itself highly discourages and forbids sudo, so being able to install Cheat without sudo access via Homebrew is ideal._

My hope would be that this Pull Request can be merged, which could then allow [Homebrew Core to have its Cheat formula updated](https://github.com/Homebrew/homebrew-core/pull/37081) to the latest version.

Presently, installing Cheat via Homebrew gets v2.3.1 (October 2018).

**Comment by m3nu** [comment]
Deploying any Python app is a pain to be honest. Works only for those who have an existing setup.

**Comment by m3nu** [comment]
I had a few hours and re-implemented `cheat` in Go with a focus on being compatible with the Python version. Partly as practice exercise. Not fully done, but maybe someone finds it useful @ryanjbonnell:

https://github.com/m3nu/go-cheat (binaries [here](https://github.com/m3nu/go-cheat/releases))

The existing cheats are packaged in the binary (only adds 150kb). No extra files to manage.

I've seen another project that fetches existing cheats from online sources, including this repo. That's another way. Since I mainly use my own cheatsheets not a big issue.

**Comment by chenrui333** [comment]
Any updates on the PR merge?

**Comment by chenrui333** [comment]
@chrisallenlane Any timeline for the new release? I am waiting for the change in the homebrew PR, https://github.com/Homebrew/homebrew-core/pull/40288 😄

**Comment by chrisallenlane** [comment]
@chenrui333, sorry for the continued delays.

I should be able to cut a release this weekend. This will likely be one of the final `cheat` Python releases.

Per #470, I'm currently re-implemented `cheat` in Golang. I'm developing it in a private repository at the moment, but will publish it when it reaches a satisfactory state. After that, with luck, all of this installation complexity should be behind us :crossed_fingers:

**Comment by chrisallenlane** [comment]
@m3nu, ha - I'm just now seeing your remarks. (Sorry!)

Yeah, per #470, I'm going to re-release an improved version of `cheat` in the near-term, which will be written in Golang. Go is definitely the way forward for this project. I feel like we're just playing whack-a-mole with these Python installation problems, and I'm tired of it.

**Pull request merged**